content/en/docs/advanced-tasks/catalog-update-formulary.md

```diff
@@ -0,0 +1,382 @@
+### Substituting a bundle version in the upgrade graph
+
+#### semver
+For all channels, replace instances of `quay.io/organization/testoperator:v1.1.0` with `quay.io/organization/testoperator:v1.1.0-CVE`
+
+```bash
+yq '(..| select(has("image") and .image == "quay.io/organization/testoperator:v1.1.0")).image = "quay.io/organization/testoperator:v1.1.0-cve"' semver.yaml
+```
+produces updated template:
+
+```yaml
+schema: olm.semver
+generatemajorchannels: false
+generateminorchannels: true
+candidate:
+  bundles:
+    - image: quay.io/organization/testoperator:v1.0.0
+    - image: quay.io/organization/testoperator:v1.0.1
+    - image: quay.io/organization/testoperator:v1.1.0-cve
+fast:
+  bundles:
+    - image: quay.io/organization/testoperator:v1.0.1
+    - image: quay.io/organization/testoperator:v1.1.0-cve
+stable:
+  bundles:
+    - image: quay.io/organization/testoperator:v1.0.1
+```
+
+#### FBC
+For all graph edges, replaces instances of `testoperator.v1.1.0` with a different bundle version `testoperator.v1.1.0-CVE`
+
+```bash
+yq '(.. | select(has("entries") and .entries[].name == "testoperator.v1.1.0" ).entries[]).name = "testoperator.v1.1.0-cve"' fbc.yaml
+```
+
+produces updated channels:
+
+```yaml
+---
+defaultChannel: stable-v1.0
+name: testoperator
+schema: olm.package
+---
+entries:
+  - name: testoperator.v1.0.0
+  - name: testoperator.v1.0.1
+    skips:
+      - testoperator.v1.0.0
+name: candidate-v1.0
+package: testoperator
+schema: olm.channel
+---
+entries:
+  - name: testoperator.v1.1.0-cve
+    replaces: testoperator.v1.0.1
+    skips:
+      - testoperator.v1.0.0
+name: candidate-v1.1
+package: testoperator
+schema: olm.channel
+---
+entries:
+  - name: testoperator.v1.0.1
+name: fast-v1.0
+package: testoperator
+schema: olm.channel
+---
+entries:
+  - name: testoperator.v1.1.0-cve
+    replaces: testoperator.v1.0.1
+name: fast-v1.1
+package: testoperator
+schema: olm.channel
+---
+entries:
+  - name: testoperator.v1.0.1
+name: stable-v1.0
+package: testoperator
+schema: olm.channel
+```
+
+### Introducing a new replacement relationship in the upgrade graph
+#### FBC
+Substitute an existing 'replaces' link target for `testoperator.v1.1.0` with a different bundle version `testoperator.v1.0.1`.
+
+
+```bash
+yq eval 'select(.schema == "olm.channel" and .name == "candidate-v1.1").entries |= [{"name" : "testoperator:v1.1.0", "replaces": "testoperator:v1.0.0"}]' fbc.yaml
+```
+ produces updated `candidate-v1.1` channel:
+
+```yaml
+...
+entries:
+  - name: testoperator:v1.1.0
+    replaces: testoperator:v1.0.0
+name: candidate-v1.1
+package: testoperator
+schema: olm.channel
+...
+```
+
+### Adding a new property to an `olm.bundle`
+
+#### FBC
+Creating a new property associated with a FBC bundle.  This example marks a bundle as deprecated to prevent installation. 
+
+```bash
+yq eval 'select(.name == "testoperator.v1.1.0" and .schema == "olm.bundle").properties += [{"type" : "olm.deprecated", "value" : "true"}]' bundle.yaml
```

Comment:
Can we pick a different property type for this example? The `olm.deprecated` property has lots of baggage. It's not really a thing in FBC, it doesn't mean what it says, and it may not even behave the same way when served from an FBC catalog.

Reply:
On second thought, I don't think it's a good idea to suggest mutating bundle properties directly in the FBC. These are typically derived from the bundle image, and I think if we want to suggest mutating bundle properties, we need to be telling folks to mutate the actual bundle, push a new bundle image with that change, and the render that bundle into the FBC.

We need to have a separate conversation (and I know @bentito has looked into this) about bundle immutability and how we can put mutable bundle-adjacent metadata in the catalog (perhaps in a different blob type)

For now, could we remove the bundle mutation examples, or change them to mutate packages or channels?

Reply:
@joelanford we need to be able to change things about a bundle without rebuilding it. Rebuilding a bundle is costly and time consuming (with additional concerns / constraints for internal release pipelines). I agree that deprecated property doesn't work the same way in FBC as it does in sqlite because of the lack of truncation in FBC but bundles with that property will still not be installable from an FBC catalog because the resolver will prevent it so it's still a good property to use and a good example of modifying a property without rebuilding a bundle IMO.

Reply:
> we need to be able to change things about a bundle without rebuilding it.

Why? Are there examples of other packaging systems where it is encouraged to:
1. Build a catalog by deriving package metadata from the packages themselves (e.g. `createrepo <pileOfRPMs>`)
2. Then tell someone to go and manipulate the index/catalog files that are generated.

I would be extremely surprised if that were the case. There are all sorts of reasons why its a bad idea:
1. Users expect the catalog not to "lie" about what the metadata of a particular package is.
2. It opens the door to all sorts of non-deterministic behavior if a package was installed using the previous version of the metadata. Consider:
   - A user stands up k8s 1.25
   - A user installs a catalog where "v1.0.0 depends on k8s >=1.25"
   - A user says "I want v1.0.0", OLM installs it
   - A new catalog shows up where 1.0.0's metadata has been updated to say "v1.0.0 depends on k8s >=1.26". What happens?
   - OR A new catalog shows up where 1.0.0's metadata has been changed to say "v1.0.0 is actually is other bundle image". What happens?

If there's a desire to have mutable bundle-adjacent metadata, we need to come up with a place to put it that keeps it separate from the bundle-derived `olm.bundle` blobs.

Reply:
To be totally honest. for arguments that amount to "but its hard/cumbersome to build push new bundles into a catalog", I think we need to address _that_ issue, not try to workaround it by compromising the integrity of the catalog.

Reply:
Would this apply to the `relatedImages` object as well?  We remove some relatedImage entries when we know we don't need them.... We have operand metadata that we use to capture which images aren't necessary and filter out uneeded images from the catalog (for mirroring performance in fan-out operators).

I think one of the core tenants is that a bundle VERSION is really immutable once a catalog is published (mutating it in development is fine).  The VERSION and the IMAGE that it references should both be immutable.  (except for relatedImages in my case).

So you don't only need to publish a new image, but you also have to add a new operator bundle version, which means that all subscribers will need to pivot to the new version. I don't disagree that it "should" be a cheap thing to strive for, but it can definitely lead to overhead if you don't understand this up front.  Hence the desire to have candidate/fast/stable channels can at least shield consumers from frequent updates.

Reply:
@cdjohnson we were also discussing these interactions in k8s slack: https://kubernetes.slack.com/archives/C0181L6JYQ2/p1685044261858409

Reply:
The consensus there was to pull the bundle-property manipulation from this doc while that conversation continues, and we can always iterate.